README
BurgerSpace - A hamburger-smashing video game.

Copyright (C) 2001-2013 Pierre Sarrazin <http://sarrazip.com/>
This program is free software distributed under the GNU GPL.
See the file COPYING.

INSTALLATION

    The program requires my `flatzebra' library and the Simple
    DirectMedia Layer multimedia library, version 1.2.0 or later
    (see http://www.libsdl.org/).  Links to both are offered on the
    BurgerSpace Home Page at

        http://sarrazip.com/dev/burgerspace.html

    See INSTALL for generic installation instructions.

    This package installs a manual page: burgerspace(6).



NEWS
2013-02-25: Version 1.9.2 released

    Slot selection when loading or saving games can now be done
    with up and down keys, to accommodate keyboard layouts where
    digits are shifted.
    Fixed bug where program crashed when saving game where
    player sprite was absent. Saved format version changed,
    but version 1 format can still be loaded.
    Added #include <unistd.h> to src/RemoteServer.cpp to resolve
    a compilation error related to close(1) under GCC 4.7.
    Fixed default UDP port displayed by ./configure --help.

2012-07-14: Version 1.9.1 released

    Now links with version 0.1.6 of flatzebra library.
    Fixed bug where chef was still displayed when in demo mode.
    Fixed compilation warnings given by gcc's -Wextra switch.
    Escaped "--" sequence in manual pages.
    Small fixes to compile under Visual C++ on Win32.
    RPM spec file rewritten using Fedora template.

2010-05-16: Version 1.9.0 released

    Two player, client-server version available.

    F11 key now toggles between full-screen mode and window mode.

    Support for XBox 360 USB Controller.

    Now requires flatzebra 0.1.5.

    Fixed compilation warnings to satisfy gcc's -Weffc++ option.

2010-03-30: Version 1.8.4 released

    Game now pauses automatically when window loses focus.
    Fixed compilation warnings.

2009-02-26: Version 1.8.3 released

    Pressing Escape key to quit now brings up a confirmation dialog.
    Various fixes to compile under GCC 4.3.
    The XPM files now define "const" character arrays.

2007-12-22: Version 1.8.2 released

    Games can be saved and loaded.  This code is experimental.

    More corner turn tolerance.  (See CORNER_TURN_TOLERANCE in
    src/BurgerSpaceEngine.cpp.)

    Updated the GPLv2 `COPYING' file to reflect new FSF address.

2004-05-02: Version 1.8.1 released

    Links with flatzebra 0.1.1.

2003-06-02: Version 1.8.0 released

    Adapted to the flatzebra library, the successor of gengameng.
    New option --z-for-pepper to use the Z key to shoot pepper
    instead of the Ctrl key.
    New .desktop file now always installed, and adapted to
    Desktop Menu Specification.
    Autoconf macro used to avoid GNU getopt code when compiling on
    BSD-like systems.
    When the sound device is not available, the program does not quit.
    When a game ends, the last level played remains displayed.

2003-01-22: Version 1.7.1 released

    When requested direction is impossible, last moving direction is
    attempted.  This can be useful to turn corners automatically.

2003-01-11: Version 1.7.0 released

    Disappointingly stupid sound effects added.
    Burger ingredients fall twice as fast.
    Makefiles cleaned up.

2002-04-03: Version 1.6.1 released

    Adaptation to gengameng 4.1.
    Contact email address removed from sources.

2001-11-25: Version 1.6 released

    Compatible with gengameng 4.0.

    Superfluous inclusion of <X11/keysym.h> now eliminated.

    ./configure script now has new option --with-gnome-desktop
    entry (or --without-gnome-desktop-entry) to decide if the
    installation adds a GNOME .desktop file and a PNG icon file to
    make BurgerSpace appear in the GNOME Games menu.  Default is
    to add the files.

    ./configure script also has new option --with-cxx to specify
    the executable program to use as the C++ compiler.  Useful
    when two versions of g++ (e.g., 2.95.2 and 3.0) cohabitate
    on a system.

2001-04-27: Version 1.5 released

    This version uses gengameng 3.0, which is based on SDL
    (http://www.libsdl.org/).  A Windows version is available.

2001-03-30: Version 1.4 released

    The signature of the onTimeOut() function in src/xtmain.cc
    has been corrected so that the file compiles on a RedHat 7.0
    system.  It should not have compiled on my RedHat 6.2 system,
    because the type of the function was not compatible in
    principle with XtAppAddTimeOut().  Strangely, NeedFunctionPrototypes
    is defined when X11/Intrinsic.h is read by g++ (2.95.2 in my case),
    so the compiler should have seen the type mismatch.
    RedHat 7.0 uses g++ 2.96, so perhaps that makes the difference.

    The FocusIn and FocusOut events are now used to turn the keyboard
    auto-repeat off and on.  Version 1.3 of BurgerSpace used the
    EnterNotify and LeaveNotify (EnterWindowMask and LeaveWindowMask)
    events instead of FocusIn and FocusOut (FocusChangeMask).
    This caused inappropriate behavior.

2001-03-23: Version 1.3 released

    The XtNinput resource is explicitly set to true in the call
    to XtVaAppInitialize() in src/xtmain.cc.  This makes sure that
    the application request the focus from the window manager.

    Thanks to David Nusinow for this patch.

2001-01-13: Version 1.2 released

    Version 1.2 uses gengameng 2.0, which avoids polling the keyboard
    with XQueryKeymap() and expects instead to be informed of key
    state changes upon KeyPress and KeyRelease events.  See xtmain.cc
    for an example of how to get those events and how to transmit
    the information to the game engine.

    This change seems to solve some of the problems of people who
    reported that the game did not receive any keyboard input.
    However, there are still reports of similar problems.



src/client.cpp
/*  $Id: client.cpp,v 1.5 2010/11/18 05:35:02 sarrazip Exp $
    client.cpp - main() function for BurgerSpace

    burgerspace - A hamburger-smashing video game.
    Copyright (C) 2001-2010 Pierre Sarrazin <http://sarrazip.com/>

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
    02110-1301, USA.
*/

#include "BurgerSpaceClient.h"

#include "LocalServer.h"
#include "RemoteServer.h"
#include "util.h"

#include <errno.h>
#include <time.h>
#include <string.h>

#include <sys/types.h>

#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#endif

#ifndef DEFAULT_UDP_SERVER_PORT
#error DEFAULT_UDP_SERVER_PORT must be defined to valid UDP port number
#endif

using namespace std;


#ifdef HAVE_GETOPT_LONG
static struct option knownOptions[] =
{
    { "help",          no_argument,       NULL, 'h' },
    { "version",       no_argument,       NULL, 'v' },
    { "initial-level", required_argument, NULL, 'i' },
    { "ms-per-frame",  required_argument, NULL, 'm' },
    { "no-sound",      no_argument,       NULL, 'n' },
    { "full-screen",   no_argument,       NULL, 'f' },
    { "old-motion",    no_argument,       NULL, 'o' },
    { "z-for-pepper",  no_argument,       NULL, 'z' },
    { "server",        required_argument, NULL, 's' },
    { "port",          required_argument, NULL, 'p' },
    { "chef",          no_argument,       NULL, 'c' },
    { "enemy",         no_argument,       NULL, 'e' },

    { NULL, 0, NULL, 0 }  // marks the end
};


static
void
displayVersionNo()
{
    cout << PROGRAM << " " << VERSION << endl;
}


static
void
displayHelp()
{
    cout << "\n";

    displayVersionNo();

    cout <<
"\n"
"Copyright (C) 2001-2010 Pierre Sarrazin <http://sarrazip.com/>\n"
"This program is free software; you may redistribute it under the terms of\n"
"the GNU General Public License.  This program has absolutely no warranty.\n"
    ;

    cout <<
"\n"
"Known options:\n"
"--help             Display this help page and exit\n"
"--version          Display this program's version number and exit\n"
"--initial-level=N  Start game at level N (N >= 1) [default=1]\n"
"--ms-per-frame=N   N milliseconds per animation frame [default=55]\n"
"                   Min. 1, max. 1000.  50 ms means 20 frames per second\n"
"--no-sound         Disable sound effects [default is to enable them]\n"
"--full-screen      Attempt to use full screen mode [default is window mode]\n"
"--old-motion       Use the old player motion system [default is new system]\n"
"--z-for-pepper     Use Z key to shoot pepper [default is Ctrl key]\n"
"--server=HOSTNAME  Run as a network client and use specified server\n"
"--port=PORT        Specify UDP port of server (only relevant with --server)\n"
"--chef             In network game, play the role of the chef\n"
"--enemy            In network game, play the role of one of the enemies\n"
"\n"
    ;
}
#endif  /* HAVE_GETOPT_LONG */


static RemoteServer *
createRemoteServer(BurgerSpaceClient &client,
                   const string &serverHostname,
                   unsigned short port,
                   Role role,
                   int &errorCode)
{
    // Try to resolve the server hostname:
    hostent *ent = gethostbyname(serverHostname.c_str());
    if (ent == NULL)
    {
        errorCode = -1;
        return NULL;
    }

    try
    {
        // Create and initialize an object representing the remote server:
        RemoteServer *rs = new RemoteServer(client,
                                            (unsigned char *) ent->h_addr_list[0],
                                            port,
                                            role);
        rs->describeClient();
        rs->requestLevelDescription();
        errorCode = 0;
        return rs;
    }
    catch (int errNo)
    {
        errorCode = errNo;
        return NULL;
    }
}


int
main(int argc, char *argv[])
{
    /*  Default game parameters:
    */
    int initLevelNo = 1;
    int millisecondsPerFrame = 55;
    bool useSound = true;
    bool fullScreen = false;
    bool useOldMotionMode = false;
    SDLKey pepperKey = SDLK_LCTRL;
    string serverHostname;  // UDP server hostname if not empty; local server otherwise
    unsigned short port = DEFAULT_UDP_SERVER_PORT;
    Role role = ROLE_SPECTATOR;


    #ifdef HAVE_GETOPT_LONG

    /*  Interpret the command-line options:
    */
    int c;
    do
    {
        c = getopt_long(argc, argv, "", knownOptions, NULL);

        switch (c)
        {
            case EOF:
                break;  // nothing to do

            case 'i':
                {
                    errno = 0;
                    long n = strtol(optarg, NULL, 10);
                    if (errno == ERANGE || n < 1 || n > 500)
                    {
                        cout << "Invalid initial level number.\n";
                        displayHelp();
                        return EXIT_FAILURE;
                    }

                    initLevelNo = int(n);
                }
                break;

            case 'm':
                {
                    errno = 0;
                    long n = strtol(optarg, NULL, 10);
                    if (errno == ERANGE || n < 1 || n > 1000)
                    {
                        cout << "Invalid number of ms per frame.\n";
                        displayHelp();
                        return EXIT_FAILURE;
                    }

                    millisecondsPerFrame = int(n);
                }
                break;

            case 'n':
                useSound = false;
                break;

            case 'f':
                fullScreen = true;
                break;

            case 'o':
                useOldMotionMode = true;
                break;

            case 'z':
                pepperKey = SDLK_z;
                break;

            case 's':
                serverHostname = optarg;
                break;

            case 'p':
                {
                    char *end = NULL;
                    long n = strtol(optarg, &end, 10);
                    if (n < 0 || n > 65535 || end == optarg || *end != '\0')
                    {
                        cerr << PROGRAM << ": --port: invalid UDP port number " << optarg << endl;
                        return EXIT_FAILURE;
                    }
                    port = static_cast<unsigned short>(n);
                }
                break;

            case 'c':
                role = ROLE_CHEF;
                break;

            case 'e':
                role = ROLE_ENEMY;
                break;

            case 'v':
                displayVersionNo();
                return EXIT_SUCCESS;

            case 'h':
                displayHelp();
                return EXIT_SUCCESS;

            default:
                displayHelp();
                return EXIT_FAILURE;
        }
    } while (c != EOF && c != '?');

    #elif defined(_MSC_VER)

    if (argc >= 2)
    {
        if (strcmp(argv[1], "--help") == 0)
        {
            cout << PROGRAM << " " << VERSION << ": copyrighted and distributed under GNU GPL\n"
                 << "\n"
                 << "Stand-alone:    " PACKAGE "\n"
                 << "Network client: " PACKAGE " SERVER [chef|enemy]\n";
            return EXIT_SUCCESS;
        }

        // argv[1] expected to be of form "SERVER KEYWORD" where
        // SERVER is a hostname and KEYWORD is either 'chef' or 'enemy'
        size_t i;
        const char *arg = argv[1];
        for (i = 0; arg[i] != '\0' && !isspace(arg[i]); ++i)
            ;

        serverHostname.assign(arg, i);  // prefix of 'arg' up to space or end

        for ( ; isspace(arg[i]); ++i)
            ;

        if (strcmp(arg + i, "chef") == 0)
            role = ROLE_CHEF;
        else if (strcmp(arg + i, "enemy") == 0)
            role = ROLE_ENEMY;
    }

    #endif  /* defined(_MSC_VER) */


    try
    {
        bool standAlone = serverHostname.empty();

        BurgerSpaceClient theBurgerSpaceClient(
                                standAlone ? ROLE_CHEF : role,
                                !standAlone /* show role */,
                                useSound, pepperKey, fullScreen,
                                standAlone);

        BurgerSpaceServerInterface *serverInterface = NULL;

        if (standAlone)
        {
            // Create a "local server":
            LocalServer *ls = new LocalServer(theBurgerSpaceClient, initLevelNo, useOldMotionMode);
            ls->finishInit();
            serverInterface = ls;
        }
        else  // remote server version:
        {
            // Upon a successful connection, createRemoteServer() sends
            // a description of this client to the server and requests
            // a level update from the server.
            //
            int errorCode = 0;
            serverInterface = createRemoteServer(theBurgerSpaceClient, serverHostname, port, role, errorCode);
            if (serverInterface == NULL)
            {
                switch (errorCode)
                {
                    case -1:
                        cerr << PROGRAM << ": failed to resolve hostname "
                                        << serverHostname << endl;
                        break;
                    default:
                        assert(errorCode > 0);
                        cerr << PROGRAM << ": failed to initialize UDP socket for server "
                                        << serverHostname << ":" << port << ": "
                                        << strerror(errorCode) << endl;
                }
                return EXIT_FAILURE;
            }

            cout << PROGRAM << ": using server " << serverHostname << " on port " << port << endl;
        }

        // Associate the created server object with the client,
        // and run the game:
        theBurgerSpaceClient.setServer(serverInterface);
        theBurgerSpaceClient.runClientMode(millisecondsPerFrame);
        serverInterface->disconnect();  // in remote mode, this notifies server of client's departure

        delete serverInterface;
    }
    catch (const string &s)
    {
        cerr << PROGRAM << ": server init error: " << s << endl;
        return EXIT_FAILURE;
    }
    catch (int e)
    {
        cerr << PROGRAM << ": init error # " << e << endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}


#ifdef _MSC_VER
int CALLBACK WinMain(__in HINSTANCE hInstance,
                     __in HINSTANCE hPrevInstance,
                     __in LPSTR lpCmdLine,
                     __in int nCmdShow)
{
    char *argv[] = { PROGRAM, NULL };
    return main(1, argv);
}
#endif  // def _MSC_VER
